File: app/src/main/kotlin/tech/mmarca/openvitals/domain/query/BodyPeriodData.kt
package tech.mmarca.openvitals.domain.query

import tech.mmarca.openvitals.domain.model.BodyFatEntry
import tech.mmarca.openvitals.domain.model.BodyWaterMassEntry
import tech.mmarca.openvitals.domain.model.BmrEntry
import tech.mmarca.openvitals.domain.model.BoneMassEntry
import tech.mmarca.openvitals.domain.model.HeightEntry
import tech.mmarca.openvitals.domain.model.LeanBodyMassEntry
import tech.mmarca.openvitals.domain.model.WeightEntry

data class BodyPeriodData(
    val weightEntries: List<WeightEntry> = emptyList(),
    val previousWeightEntries: List<WeightEntry> = emptyList(),
    val baselineWeightEntries: List<WeightEntry> = emptyList(),
    val latestWeightKg: Double? = null,
    val heightCm: Double? = null,
    val heightEntries: List<HeightEntry> = emptyList(),
    val previousHeightEntries: List<HeightEntry> = emptyList(),
    val baselineHeightEntries: List<HeightEntry> = emptyList(),
    val bodyFatEntries: List<BodyFatEntry> = emptyList(),
    val previousBodyFatEntries: List<BodyFatEntry> = emptyList(),
    val baselineBodyFatEntries: List<BodyFatEntry> = emptyList(),
    val latestBodyFatPercent: Double? = null,
    val leanMassKg: Double? = null,
    val leanMassEntries: List<LeanBodyMassEntry> = emptyList(),
    val previousLeanMassEntries: List<LeanBodyMassEntry> = emptyList(),
    val baselineLeanMassEntries: List<LeanBodyMassEntry> = emptyList(),
    val bmrKcal: Double? = null,
    val bmrEntries: List<BmrEntry> = emptyList(),
    val previousBmrEntries: List<BmrEntry> = emptyList(),
    val baselineBmrEntries: List<BmrEntry> = emptyList(),
    val boneMassKg: Double? = null,
    val boneMassEntries: List<BoneMassEntry> = emptyList(),
    val previousBoneMassEntries: List<BoneMassEntry> = emptyList(),
    val baselineBoneMassEntries: List<BoneMassEntry> = emptyList(),
    val bodyWaterMassKg: Double? = null,
    val bodyWaterMassEntries: List<BodyWaterMassEntry> = emptyList(),
    val previousBodyWaterMassEntries: List<BodyWaterMassEntry> = emptyList(),
    val baselineBodyWaterMassEntries: List<BodyWaterMassEntry> = emptyList(),
)

File: lib/domain/query/body_period_data.dart
import 'package:freezed_annotation/freezed_annotation.dart';

import '../model/body_models.dart';

part 'body_period_data.freezed.dart';

@freezed
abstract class BodyPeriodData with _$BodyPeriodData {
  const factory BodyPeriodData({
    @Default(<WeightEntry>[]) List<WeightEntry> weightEntries,
    @Default(<WeightEntry>[]) List<WeightEntry> previousWeightEntries,
    @Default(<WeightEntry>[]) List<WeightEntry> baselineWeightEntries,
    double? latestWeightKg,
    double? heightCm,
    @Default(<HeightEntry>[]) List<HeightEntry> heightEntries,
    @Default(<HeightEntry>[]) List<HeightEntry> previousHeightEntries,
    @Default(<HeightEntry>[]) List<HeightEntry> baselineHeightEntries,
    @Default(<BodyFatEntry>[]) List<BodyFatEntry> bodyFatEntries,
    @Default(<BodyFatEntry>[]) List<BodyFatEntry> previousBodyFatEntries,
    @Default(<BodyFatEntry>[]) List<BodyFatEntry> baselineBodyFatEntries,
    double? latestBodyFatPercent,
    double? leanMassKg,
    @Default(<LeanBodyMassEntry>[]) List<LeanBodyMassEntry> leanMassEntries,
    @Default(<LeanBodyMassEntry>[])
    List<LeanBodyMassEntry> previousLeanMassEntries,
    @Default(<LeanBodyMassEntry>[])
    List<LeanBodyMassEntry> baselineLeanMassEntries,
    double? bmrKcal,
    @Default(<BmrEntry>[]) List<BmrEntry> bmrEntries,
    @Default(<BmrEntry>[]) List<BmrEntry> previousBmrEntries,
    @Default(<BmrEntry>[]) List<BmrEntry> baselineBmrEntries,
    double? boneMassKg,
    @Default(<BoneMassEntry>[]) List<BoneMassEntry> boneMassEntries,
    @Default(<BoneMassEntry>[]) List<BoneMassEntry> previousBoneMassEntries,
    @Default(<BoneMassEntry>[]) List<BoneMassEntry> baselineBoneMassEntries,
    double? bodyWaterMassKg,
    @Default(<BodyWaterMassEntry>[])
    List<BodyWaterMassEntry> bodyWaterMassEntries,
    @Default(<BodyWaterMassEntry>[])
    List<BodyWaterMassEntry> previousBodyWaterMassEntries,
    @Default(<BodyWaterMassEntry>[])
    List<BodyWaterMassEntry> baselineBodyWaterMassEntries,
  }) = _BodyPeriodData;
}
